client.go

```diff
@@ -151,6 +152,10 @@ func (client *client) Close() error {
 		return ErrClosedClient
 	}
 
+	// shutdown and wait for the background thread before we take the lock, to avoid races
+	close(client.closer)
+	<-client.closed
```

Comment:
It will, however, deadlock if the client calls `Close()` on itself because of an error during initialization. So something still to fix.